File: POUCH_APP/frontend/src/domain/israeliId.ts
/**
 * Israeli teudat zehut check digit. Mirrors backend/app/core/israeli_id.py.
 *
 * Random 9-digit strings fail this ~90% of the time — which is why the national ID
 * is optional and validated, and the internal MRN is the key.
 * Client-side validation is a convenience only; the server validates again.
 */

/**
 * Common separators people type into an ID field. Anything else — letters in
 * particular — makes the input invalid rather than being silently stripped:
 * "abcdefghij0" must not reduce to "0" and then validate as 000000000.
 */
const SEPARATORS = /[\s\-.]/g;
const DIGITS_ONLY = /^\d{1,9}$/;

export function normaliseNationalId(value: string | null | undefined): string | null {
  if (!value) return null;
  const candidate = value.replace(SEPARATORS, '');
  if (!DIGITS_ONLY.test(candidate)) return null;
  return candidate.padStart(9, '0');
}

export function isValidIsraeliId(value: string | null | undefined): boolean {
  const digits = normaliseNationalId(value);
  if (digits === null) return false;

  let total = 0;
  for (let index = 0; index < 9; index += 1) {
    const doubled = Number(digits[index]) * (index % 2 === 0 ? 1 : 2);
    total += doubled < 10 ? doubled : doubled - 9;
  }
  return total % 10 === 0;
}

export function maskNationalId(value: string | null): string {
  if (!value) return '—';
  return `•••••${value.slice(-4)}`;
}

File: POUCH_APP/backend/app/core/israeli_id.py
"""Israeli teudat zehut check-digit validation.

Mirrors frontend/src/domain/israeliId.ts.

Random 9-digit strings fail this ~90% of the time, which is why the national ID is an
optional, validated field and never the primary key -- the internal MRN is.
"""

from __future__ import annotations

import re

#: Common separators people type into an ID field. Anything else -- letters in
#: particular -- makes the input invalid rather than being silently stripped:
#: "abcdefghij0" must not reduce to "0" and then validate as 000000000.
_SEPARATORS = re.compile(r"[\s\-.]")
_DIGITS_ONLY = re.compile(r"^\d{1,9}$")
_EIGHT_DIGITS = re.compile(r"^\d{8}$")


def normalise(value: str | None) -> str | None:
    """Strip separators and left-pad to 9 digits, or None if not a plausible ID."""
    if not value:
        return None
    candidate = _SEPARATORS.sub("", value)
    if not _DIGITS_ONLY.match(candidate):
        return None
    return candidate.zfill(9)


def is_valid_israeli_id(value: str | None) -> bool:
    digits = normalise(value)
    if digits is None:
        return False

    total = 0
    for index, char in enumerate(digits):
        doubled = int(char) * (1 if index % 2 == 0 else 2)
        total += doubled if doubled < 10 else doubled - 9
    return total % 10 == 0


def check_digit(first_eight: str) -> int:
    """Compute the 9th digit for an 8-digit stem -- useful for generating test IDs."""
    stem = _SEPARATORS.sub("", first_eight).zfill(8)
    if not _EIGHT_DIGITS.match(stem):
        raise ValueError(f"expected 8 digits, got {first_eight!r}")

    total = 0
    for index, char in enumerate(stem):
        doubled = int(char) * (1 if index % 2 == 0 else 2)
        total += doubled if doubled < 10 else doubled - 9
    return (10 - total % 10) % 10
